src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+const std = @import("std");
+const GenCatData = @import("GenCatData");
+
+/// The token kinds
+pub const T = enum(u8) {
+    EndOfFile,
+
+    // primitives
+    Float,
+    String,
+    SingleQuote,
+
+    // a part of a string interpolation; generally you'll see something like:
+    // StringBegin, OpenCurly, <expr>, CloseCurly StringPart, OpenCurly, <expr>, CloseCurly
+    StringBegin,
+    StringPart,
+
+    // These are not technically valid, but we can have the formatter fix them up.
+    SingleQuoteBegin,
+    SingleQuotePart,
+
+    UpperIdent,
+    LowerIdent,
+    Underscore,
+    DotLowerIdent,
+    DotNumber,
+
+    OpenRound,
+    CloseRound,
+    OpenSquare,
+    CloseSquare,
+    OpenCurly,
+    CloseCurly,
+
+    OpPlus,
+    OpStar,
+    OpPizza,
+    OpAssign,
+    OpBinaryMinus, // trailing whitespace
+    OpUnaryMinus, // no trailing whitespace
+    OpNotEquals,
+    OpBang,
+    OpAnd,
+    OpAmpersand,
+    OpQuestion,
+    OpOr,
+    OpBar,
+    OpDoubleSlash,
+    OpSlash,
+    OpPercent,
+    OpCaret,
+    OpGreaterThanOrEq,
+    OpGreaterThan,
+    OpLessThanOrEq,
+    OpBackArrow,
+    OpLessThan,
+    OpEquals,
+    OpColonEqual,
+
+    Comma,
+    Dot,
+    DoubleDot,
+    TripleDot,
+    OpColon,
+    OpArrow,
+    OpBackslash,
+
+    // Keywords
+    KwIf,
+    KwThen,
+    KwElse,
+    KwWhen,
+    KwIs,
+    KwAs,
+    KwDbg,
+    KwCrash,
+    KwHas,
+    KwWhere,
+    KwImplements,
+    KwExposes,
+    KwImports,
+    KwImport,
+    KwWith,
+    KwGenerates,
+    KwPackage,
+    KwPackages,
+    KwRequires,
+    KwProvides,
+    KwTo,
+    KwInterface,
+    KwApp,
+    KwPlatform,
+    KwHosted,
+    KwDebug,
+    KwExpect,
+
+    NamedUnderscore,
+    OpaqueName,
+    Int,
+    DotUpperIdent,
+    NoSpaceDotNumber,
+    NoSpaceDotLowerIdent,
+    NoSpaceDotUpperIdent,
+};
+
+/// A helper function equivalent to the Rust impl T { is_keyword }.
+pub fn isKeyword(tok: T) bool {
+    return switch (tok) {
+        .KwIf,
+        .KwThen,
+        .KwElse,
+        .KwWhen,
+        .KwIs,
+        .KwAs,
+        .KwDbg,
+        .KwExpect,
+        .KwCrash,
+        .KwHas,
+        .KwWhere,
+        .KwImplements,
+        .KwExposes,
+        .KwImport,
+        .KwImports,
+        .KwWith,
+        .KwGenerates,
+        .KwPackage,
+        .KwPackages,
+        .KwRequires,
+        .KwProvides,
+        .KwTo,
+        .KwInterface,
+        .KwApp,
+        .KwPlatform,
+        .KwHosted,
+        .KwExpectFx,
+        => true,
+        else => false,
+    };
+}
+
+pub const Token = struct {
+    kind: T,
+    offset: usize,
+    length: usize,
```

Comment:
I assume this could be a function instead of wasting memory? Though if this includes full strings as a single token then maybe we want to store the length, just in a small type, maybe u16 or something.

Reply:
Are you referring to length, specifically?

My long-term plan is indeed to compute that in a function, but: (1) it's easy to record this now and use that to verify the length computation later, and (2) that length computation can get out-of-sync with the main tokenizer. That's one divergence I'd like to avoid for now by simply making it impossible.

Reply:
Yeah, was referring to length specifically.